File: native/windows/Services/ProfessionalMonitor.cs
using NikonLink.Windows.Models;
using System.Windows;
using System.Windows.Media;
using System.Windows.Media.Imaging;

namespace NikonLink.Windows.Services;

public sealed record ProfessionalMonitorResult(
    BitmapSource Image,
    string RedHistogram,
    string GreenHistogram,
    string BlueHistogram,
    string Waveform,
    string Vectorscope,
    int PeakingCoverage);

public static class ProfessionalMonitor
{
    private const int ScopeColumns = 64;
    private const int ScopeRows = 48;
    private const string Hex = "0123456789ABCDEF";

    public static ProfessionalMonitorResult Process(
        BitmapSource source,
        bool focusPeaking,
        bool falseColor) => Process(source, focusPeaking, falseColor, null);

    internal static ProfessionalMonitorResult Process(
        BitmapSource source,
        bool focusPeaking,
        bool falseColor,
        NikonCloudPreset? nikonCloudPreset)
    {
        BitmapSource working = new FormatConvertedBitmap(
            source,
            PixelFormats.Bgra32,
            null,
            0);
        var scale = Math.Min(
            1,
            640.0 / Math.Max(working.PixelWidth, working.PixelHeight));
        if (scale < 1)
        {
            var resized = new TransformedBitmap(
                working,
                new ScaleTransform(scale, scale));
            resized.Freeze();
            working = resized;
        }
        var width = working.PixelWidth;
        var height = working.PixelHeight;
        var stride = width * 4;
        var pixels = new byte[stride * height];
        working.CopyPixels(pixels, stride, 0);
        var luminance = focusPeaking
            ? new int[width * height]
            : null;
        var red = ScopeBuffer();
        var green = ScopeBuffer();
        var blue = ScopeBuffer();
        var lumaScope = ScopeBuffer();
        var cbScope = ScopeBuffer();
        var crScope = ScopeBuffer();

        for (var y = 0; y < height; y++)
        {
            for (var x = 0; x < width; x++)
            {
                var pixel = y * width + x;
                var offset = y * stride + x * 4;
                var b = (int)pixels[offset];
                var g = (int)pixels[offset + 1];
                var r = (int)pixels[offset + 2];
                if (nikonCloudPreset is not null)
                {
                    var cloudRed = r / 255.0;
                    var cloudGreen = g / 255.0;
                    var cloudBlue = b / 255.0;
                    nikonCloudPreset.ApplyPreviewEffect(
                        ref cloudRed,
                        ref cloudGreen,
                        ref cloudBlue);
                    r = Clamp8((int)Math.Round(cloudRed * 255));
                    g = Clamp8((int)Math.Round(cloudGreen * 255));
                    b = Clamp8((int)Math.Round(cloudBlue * 255));
                    pixels[offset] = (byte)b;
                    pixels[offset + 1] = (byte)g;
                    pixels[offset + 2] = (byte)r;
                    pixels[offset + 3] = 255;
                }
                var value = (54 * r + 183 * g + 19 * b) >> 8;
                if (luminance is not null)
                {
                    luminance[pixel] = value;
                }
                var column = Math.Min(ScopeColumns - 1, x * ScopeColumns / Math.Max(1, width));
                Accumulate(red, column, r);
                Accumulate(green, column, g);
                Accumulate(blue, column, b);
                Accumulate(lumaScope, column, value);
                var cb = Clamp8(128 + ((-29 * r - 99 * g + 128 * b) >> 8));
                var cr = Clamp8(128 + ((128 * r - 116 * g - 12 * b) >> 8));
                Accumulate(cbScope, column, cb);
                Accumulate(crScope, column, cr);
                if (falseColor)
                {
                    var color = FalseColor(value);
                    pixels[offset] = color.B;
                    pixels[offset + 1] = color.G;
                    pixels[offset + 2] = color.R;
                }
            }
        }

        var peakingPixels = 0;
        if (focusPeaking && width > 2 && height > 2)
        {
            for (var y = 1; y < height - 1; y++)
            {
                for (var x = 1; x < width - 1; x++)
                {
                    var pixel = y * width + x;
                    var horizontal = Math.Abs(
                        luminance![pixel - 1] - luminance[pixel + 1]);
                    var vertical = Math.Abs(
                        luminance[pixel - width] -
                        luminance[pixel + width]);
                    if (horizontal + vertical <= 72)
                    {
                        continue;
                    }
                    var offset = y * stride + x * 4;
                    pixels[offset] = 205;
                    pixels[offset + 1] = 38;
                    pixels[offset + 2] = 255;
                    pixels[offset + 3] = 255;
                    peakingPixels++;
                }
            }
        }

        BitmapSource output;
        if (focusPeaking || falseColor || nikonCloudPreset is not null)
        {
            var rendered = new WriteableBitmap(
                width,
                height,
                source.DpiX,
                source.DpiY,
                PixelFormats.Bgra32,
                null);
            rendered.WritePixels(
                new Int32Rect(0, 0, width, height),
                pixels,
                stride,
                0);
            rendered.Freeze();
            output = rendered;
        }
        else
        {
            working.Freeze();
            output = working;
        }
        return new ProfessionalMonitorResult(
            output,
            DensityMap(red),
            DensityMap(green),
            DensityMap(blue),
            DensityMap(lumaScope),
            $"{DensityMap(cbScope)}|{DensityMap(crScope)}",
            (int)Math.Round(
                peakingPixels * 100.0 / Math.Max(1, width * height)));
    }

    private static int[] ScopeBuffer() => new int[ScopeColumns * ScopeRows];

    private static void Accumulate(int[] buffer, int column, int value)
    {
        var row = ScopeRows - 1 - Clamp8(value) * (ScopeRows - 1) / 255;
        buffer[row * ScopeColumns + column]++;
    }

    private static int Clamp8(int value) => Math.Min(255, Math.Max(0, value));

    private static Color FalseColor(int luma) => luma switch
    {
        < 32 => Color.FromRgb(18, 24, 150),
        < 64 => Color.FromRgb(0, 150, 255),
        < 112 => Color.FromRgb(32, 205, 120),
        < 160 => Color.FromRgb(126, 126, 126),
        < 208 => Color.FromRgb(255, 205, 20),
        < 240 => Color.FromRgb(255, 92, 32),
        _ => Color.FromRgb(255, 32, 56)
    };

    private static string DensityMap(int[] values)
    {
        var maximum = Math.Max(1, values.Max());
        var divisor = Math.Log(1.0 + maximum);
        return $"S{ScopeColumns}x{ScopeRows}:" + new string(values.Select(value =>
            Hex[Math.Min(15, Math.Max(0,
                (int)Math.Round(Math.Log(1.0 + value) / divisor * 15)))])
            .ToArray());
    }
}

File: native/android/app/src/main/java/com/tauber/nikonlink/ProfessionalMonitor.java
package com.tauber.nikonlink;

import android.graphics.Bitmap;
import android.graphics.Color;

final class ProfessionalMonitor {
    static final class Result {
        final Bitmap image;
        final String redHistogram;
        final String greenHistogram;
        final String blueHistogram;
        final String waveform;
        final String vectorscope;
        final int peakingCoverage;

        Result(
                Bitmap image,
                String redHistogram,
                String greenHistogram,
                String blueHistogram,
                String waveform,
                String vectorscope,
                int peakingCoverage) {
            this.image = image;
            this.redHistogram = redHistogram;
            this.greenHistogram = greenHistogram;
            this.blueHistogram = blueHistogram;
            this.waveform = waveform;
            this.vectorscope = vectorscope;
            this.peakingCoverage = peakingCoverage;
        }
    }

    private static final int SCOPE_COLUMNS = 64;
    private static final int SCOPE_ROWS = 48;
    private static final char[] HEX = "0123456789ABCDEF".toCharArray();

    static Result process(
            Bitmap source,
            boolean focusPeaking,
            boolean falseColor,
            NikonCloudPreview.Preset nikonCloudPreset) {
        int sourceWidth = source.getWidth();
        int sourceHeight = source.getHeight();
        double scale = Math.min(1.0, 640.0 / Math.max(sourceWidth, sourceHeight));
        int width = Math.max(1, (int) Math.round(sourceWidth * scale));
        int height = Math.max(1, (int) Math.round(sourceHeight * scale));
        Bitmap working = scale < 1
                ? Bitmap.createScaledBitmap(source, width, height, true)
                : source;
        int[] pixels = new int[width * height];
        int[] luma = focusPeaking ? new int[width * height] : null;
        int[] red = scopeBuffer();
        int[] green = scopeBuffer();
        int[] blue = scopeBuffer();
        int[] lumaScope = scopeBuffer();
        int[] cbScope = scopeBuffer();
        int[] crScope = scopeBuffer();
        double[] cloudChannels = nikonCloudPreset == null
                ? null : new double[3];
        double[] cloudHsl = nikonCloudPreset == null
                ? null : new double[3];
        working.getPixels(pixels, 0, width, 0, 0, width, height);

        for (int y = 0; y < height; y++) {
            for (int x = 0; x < width; x++) {
                int index = y * width + x;
                int color = pixels[index];
                int r = Color.red(color);
                int g = Color.green(color);
                int b = Color.blue(color);
                if (nikonCloudPreset != null) {
                    cloudChannels[0] = r / 255.0;
                    cloudChannels[1] = g / 255.0;
                    cloudChannels[2] = b / 255.0;
                    nikonCloudPreset.applyPreviewEffect(
                            cloudChannels,
                            cloudHsl);
                    r = clamp8((int) Math.round(cloudChannels[0] * 255));
                    g = clamp8((int) Math.round(cloudChannels[1] * 255));
                    b = clamp8((int) Math.round(cloudChannels[2] * 255));
                    pixels[index] = Color.rgb(r, g, b);
                }
                int value = (54 * r + 183 * g + 19 * b) >> 8;
                if (luma != null) luma[index] = value;
                int column = Math.min(SCOPE_COLUMNS - 1, x * SCOPE_COLUMNS / Math.max(1, width));
                accumulate(red, column, r);
                accumulate(green, column, g);
                accumulate(blue, column, b);
                accumulate(lumaScope, column, value);
                int cb = clamp8(128 + ((-29 * r - 99 * g + 128 * b) >> 8));
                int cr = clamp8(128 + ((128 * r - 116 * g - 12 * b) >> 8));
                accumulate(cbScope, column, cb);
                accumulate(crScope, column, cr);
                if (falseColor) {
                    pixels[index] = falseColor(value);
                }
            }
        }

        int peakingPixels = 0;
        if (focusPeaking && width > 2 && height > 2) {
            for (int y = 1; y < height - 1; y++) {
                for (int x = 1; x < width - 1; x++) {
                    int index = y * width + x;
                    int horizontal = Math.abs(luma[index - 1] - luma[index + 1]);
                    int vertical = Math.abs(
                            luma[index - width] - luma[index + width]);
                    if (horizontal + vertical > 72) {
                        pixels[index] = Color.rgb(255, 38, 205);
                        peakingPixels++;
                    }
                }
            }
        }
        Bitmap output;
        if (focusPeaking || falseColor || nikonCloudPreset != null) {
            output = Bitmap.createBitmap(
                    pixels,
                    width,
                    height,
                    Bitmap.Config.ARGB_8888);
            if (working != source) working.recycle();
        } else {
            output = working;
        }
        return new Result(
                output,
                densityMap(red),
                densityMap(green),
                densityMap(blue),
                densityMap(lumaScope),
                densityMap(cbScope) + "|" + densityMap(crScope),
                (int) Math.round(
                        peakingPixels * 100.0 / Math.max(1, width * height)));
    }

    private static int[] scopeBuffer() {
        return new int[SCOPE_COLUMNS * SCOPE_ROWS];
    }

    private static void accumulate(int[] buffer, int column, int value) {
        int row = SCOPE_ROWS - 1 - clamp8(value) * (SCOPE_ROWS - 1) / 255;
        buffer[row * SCOPE_COLUMNS + column]++;
    }

    private static int clamp8(int value) {
        return Math.min(255, Math.max(0, value));
    }

    private static int falseColor(int luma) {
        if (luma < 32) return Color.rgb(18, 24, 150);
        if (luma < 64) return Color.rgb(0, 150, 255);
        if (luma < 112) return Color.rgb(32, 205, 120);
        if (luma < 160) return Color.rgb(126, 126, 126);
        if (luma < 208) return Color.rgb(255, 205, 20);
        if (luma < 240) return Color.rgb(255, 92, 32);
        return Color.rgb(255, 32, 56);
    }

    private static String densityMap(int[] values) {
        int maximum = 1;
        for (int value : values) maximum = Math.max(maximum, value);
        double divisor = Math.log1p(maximum);
        StringBuilder result = new StringBuilder(8 + values.length);
        result.append('S').append(SCOPE_COLUMNS).append('x').append(SCOPE_ROWS).append(':');
        for (int value : values) {
            int level = (int) Math.round(Math.log1p(value) / divisor * 15);
            result.append(HEX[Math.min(15, Math.max(0, level))]);
        }
        return result.toString();
    }
}
